First usage or "installation" issues

I'm consistently experiencing an error message when trying to use the Azure MCP Server for the first time. Whether I use the repo's button-click experience [which is SO slick, btw, love it] to install it or I use the in-editor "Add MCP Server" feature, _or_ when I call the `npx` command from the terminal, I get the same error. I've reproduced this on one machine multiple times, and tried on another machine and it worked.

**Machine Summary**

1. Machine A (Bad): Windows Desktop, Intel chipset, 64bit - OS has been installed for some time, as has been Node, npx, etc. When I first ran into this issue on the desktop machine I completely removed Node, NPX, all of it. Deleted all associated folders and environment variables from my path, rebooted, and then tried the install process via the repo button link. Got the `install npx` experience inside VS Code, downloaded it and installed it. The issue persists post-installation. 
2. Machine B (Good): Windows Laptop, Intel chipset, 64bit - repaved 5 days ago, completely fresh install of `npx` via the `install npx` command that pops up for first-time VS Code users when they click the repo button. 

When I run this command:

```
node --version
npx --version
```

On Machine A (the desktop that keeps failing), I get this: 

![Image](https://github.com/user-attachments/assets/730ab7a1-83ab-4a94-81ca-37964aa62fb3)

On Machine B (the laptop that works), I get this: 

![Image](https://github.com/user-attachments/assets/ab24b7bc-e274-4207-9ea3-731d81d7b996)

**Error Details**

On the desktop machine (Machine A, the one that isn't working), with 100% consistency across both Visual Studio Code Insiders and Visual Studio Code "normal," upon clicking the `start` button whilst in `settings.json`, I see this error message. This is the error I was seeing prior to uninstalling/re-installing Node on the desktop: 

```
2025-05-29 11:27:14.349 [info] Starting server azure_mcp_server
2025-05-29 11:27:14.350 [info] Connection state: Starting
2025-05-29 11:27:14.350 [info] Starting server from LocalProcess extension host
2025-05-29 11:27:14.362 [info] Connection state: Starting
2025-05-29 11:27:14.364 [info] Connection state: Running
2025-05-29 11:27:15.688 [warning] [server stderr] Failed to load platform specific package '@azure/mcp-win32-x64': Cannot find module '@azure/mcp-win32-x64'
2025-05-29 11:27:15.688 [warning] [server stderr] Require stack:
2025-05-29 11:27:15.688 [warning] [server stderr] - C:\Users\brady\AppData\Roaming\npm\node_modules\@azure\mcp\index.js
2025-05-29 11:27:15.708 [info] Connection state: Error Process exited with code 1
2025-05-29 11:27:15.708 [error] Server exited before responding to `initialize` request.
```

To see if this was a problem consistent with Node/npx on my machine, I also used the Playwright repo to install the Playwright MCP Server. It starts up without error: 

![Image](https://github.com/user-attachments/assets/3718a054-07b7-4776-be4c-d97c4c272089)

When I use the tools button in the Copilot window or the command palette to see the tools available to me given my various MCP Servers, I see that the Playwright one is working, whereas the Azure MCP one is indeed not working when on Machine A, the desktop: 

![Image](https://github.com/user-attachments/assets/011c90db-a84e-4317-90d5-58ba4e1b4504)

Let me know if you need any additional information, want to do a screen-share - whatever you need to help troubleshoot this and ascertain what's up. Note: I did try to use the Aspire Dashboard via a docker image per the troubleshooting guide. Nothing pops up in the Aspire Dashboard when the server tries to light up, so, I'm not sure if that's due to the fact that the server never starts, thus, there's no telemetry being streamed, or if I've configured it incorrectly. 

**.NET Version Info**

Just in case, and per the genius suggestion of @timheuer, I took a peek at the .NET versions I have on both machines. 

Machine A (the broken one): 

```
.NET SDK:
Version:           10.0.100-preview.4.25258.110
Commit:            c22dcd0c7a
Workload version:  10.0.100-manifests.c9bc7fa7
MSBuild version:   17.15.0-preview-25258-110+c22dcd0c7
```

Machine B (the one that works)

```
.NET SDK:
Version:           9.0.300
Commit:            15606fe0a8
Workload version:  9.0.300-manifests.af4147de
MSBuild version:   17.14.5+edd3bbf37
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,10 +3,11 @@
 ## 0.1.1 (Unreleased)
 
 ### Features Added
-
 ### Breaking Changes
 
 ### Bugs Fixed
+
+- Fixed return value of `tools/list` to use JSON object names. https://github.com/Azure/azure-mcp/pull/275
 
 ### Other Changes
 

--- a/src/Commands/Server/ToolOperations.cs
+++ b/src/Commands/Server/ToolOperations.cs
@@ -146,7 +146,6 @@ public class ToolOperations
 
         var options = command.GetCommand().Options;
 
-
         var schema = new JsonObject
         {
             ["type"] = "object"
@@ -159,7 +158,7 @@ public class ToolOperations
             {
                 arguments.Add(option.Name, new JsonObject()
                 {
-                    ["type"] = option.ValueType.ToString().ToLower(),
+                    ["type"] = option.ValueType.ToJsonType(),
                     ["description"] = option.Description,
                 });
             }

--- a/src/Commands/Server/TypeToJsonTypeMapper.cs
+++ b/src/Commands/Server/TypeToJsonTypeMapper.cs
@@ -69,7 +69,7 @@ public static class TypeToJsonTypeMapper
 
         if (type.IsEnum)
         {
-            return "string"; // or integer
+            return "integer";
         }
 
         return "object";

--- a/src/Commands/Server/TypeToJsonTypeMapper.cs
+++ b/src/Commands/Server/TypeToJsonTypeMapper.cs
@@ -1,0 +1,77 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections;
+
+namespace AzureMcp.Commands.Server;
+
+/// <summary>
+/// Gets the JSON object type based on its C# type.
+/// </summary>
+public static class TypeToJsonTypeMapper
+{
+    private static readonly Dictionary<Type, string> TypeToJsonMap = new()
+    {
+        // String types
+        { typeof(string), "string" },
+        { typeof(char), "string" },
+        { typeof(Guid), "string" },
+        { typeof(DateTime), "string" },
+        { typeof(DateTimeOffset), "string" },
+        { typeof(TimeSpan), "string" },
+        { typeof(Uri), "string" },
+        
+        // Number types
+        { typeof(int), "integer" },
+        { typeof(uint), "integer" },
+        { typeof(long), "integer" },
+        { typeof(ulong), "integer" },
+        { typeof(short), "integer" },
+        { typeof(ushort), "integer" },
+        { typeof(byte), "integer" },
+        { typeof(sbyte), "integer" },
+
+        { typeof(float), "number" },
+        { typeof(double), "number" },
+        { typeof(decimal), "number" },
+        
+        // Boolean
+        { typeof(bool), "boolean" },
+        
+        // Arrays and collections
+        { typeof(Array), "array" },
+        
+        // Object
+        { typeof(object), "object" }
+    };
+
+    /// <summary>
+    /// Gets the JSON type name based on the given type.  If <paramref name="type"/> is null, then "null" is returned.
+    /// </summary>
+    /// <param name="type">Type to get JSON type name from.</param>
+    /// <returns></returns>
+    public static string ToJsonType(this Type? type)
+    {
+        if (type == null)
+        {
+            return "null";
+        }
+
+        if (TypeToJsonMap.TryGetValue(type, out string? jsonType) && jsonType != null)
+        {
+            return jsonType;
+        }
+
+        if (typeof(IEnumerable).IsAssignableFrom(type) && type != typeof(string))
+        {
+            return typeof(IDictionary).IsAssignableFrom(type) ? "object" : "array";
+        }
+
+        if (type.IsEnum)
+        {
+            return "string"; // or integer
+        }
+
+        return "object";
+    }
+}

--- a/tests/Commands/Server/ToolOperationsTest.cs
+++ b/tests/Commands/Server/ToolOperationsTest.cs
@@ -1,8 +1,11 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+using System.Text.Json;
+using System.Xml.Linq;
 using AzureMcp.Commands;
 using AzureMcp.Commands.Server;
+using Microsoft.AspNetCore.Components.Forms;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Logging;
 using ModelContextProtocol.Protocol;
@@ -13,6 +16,18 @@ using Xunit;
 namespace AzureMcp.Tests.Commands.Server;
 public class ToolOperationsTest
 {
+    // https://json-schema.org/understanding-json-schema/reference/type
+    private static readonly HashSet<string> JsonSchemaDataTypes = new()
+    {
+        "string",
+        "integer",
+        "number",
+        "boolean",
+        "array",
+        "null",
+        "object"
+    };
+
     private readonly CommandFactory _commandFactory;
     private readonly IServiceProvider _serviceProvider;
     private readonly ILogger<ToolOperations> _logger;
@@ -48,6 +63,31 @@ public class ToolOperationsTest
             Assert.NotNull(tool);
             Assert.NotNull(tool.Name);
             Assert.NotNull(tool.Description!);
+
+            Assert.Equal(JsonValueKind.Object, tool.InputSchema.ValueKind);
+
+            foreach (var properties in tool.InputSchema.EnumerateObject())
+            {
+                if (properties.NameEquals("type"))
+                {
+                    Assert.Equal("object", properties.Value.GetString());
+                }
+
+                if (!properties.NameEquals("properties"))
+                {
+                    continue;
+                }
+
+                var commandArguments = properties.Value.EnumerateObject().ToArray();
+                foreach (var argument in commandArguments)
+                {
+                    var argumentType = argument.Value.GetProperty("type");
+                    var value = argumentType.GetString();
+
+                    Assert.NotNull(value);
+                    Assert.Contains(value, JsonSchemaDataTypes);
+                }
+            }
         }
     }
 }

--- a/tests/Commands/Server/ToolOperationsTest.cs
+++ b/tests/Commands/Server/ToolOperationsTest.cs
@@ -2,10 +2,8 @@
 // Licensed under the MIT License.
 
 using System.Text.Json;
-using System.Xml.Linq;
 using AzureMcp.Commands;
 using AzureMcp.Commands.Server;
-using Microsoft.AspNetCore.Components.Forms;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Logging;
 using ModelContextProtocol.Protocol;

--- a/tests/Commands/Server/TypeToJsonTypeMapperTests.cs
+++ b/tests/Commands/Server/TypeToJsonTypeMapperTests.cs
@@ -1,0 +1,258 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections;
+using AzureMcp.Commands.Server;
+using Xunit;
+
+namespace AzureMcp.Tests.Commands.Server;
+
+public class TypeToJsonTypeMapperTests
+{
+    [Fact]
+    public void ToJsonType_WithNullType_ReturnsNull()
+    {
+        // Arrange
+        Type? nullType = null;
+
+        // Act
+        var result = nullType.ToJsonType();
+
+        // Assert
+        Assert.Equal("null", result);
+    }
+
+    [Theory]
+    [InlineData(typeof(string), "string")]
+    [InlineData(typeof(char), "string")]
+    [InlineData(typeof(Guid), "string")]
+    [InlineData(typeof(DateTime), "string")]
+    [InlineData(typeof(DateTimeOffset), "string")]
+    [InlineData(typeof(TimeSpan), "string")]
+    [InlineData(typeof(Uri), "string")]
+    public void ToJsonType_WithStringTypes_ReturnsString(Type type, string expected)
+    {
+        // Act
+        var result = type.ToJsonType();
+
+        // Assert
+        Assert.Equal(expected, result);
+    }
+
+    [Theory]
+    [InlineData(typeof(int), "integer")]
+    [InlineData(typeof(uint), "integer")]
+    [InlineData(typeof(long), "integer")]
+    [InlineData(typeof(ulong), "integer")]
+    [InlineData(typeof(short), "integer")]
+    [InlineData(typeof(ushort), "integer")]
+    [InlineData(typeof(byte), "integer")]
+    [InlineData(typeof(sbyte), "integer")]
+    public void ToJsonType_WithIntegerTypes_ReturnsInteger(Type type, string expected)
+    {
+        // Act
+        var result = type.ToJsonType();
+
+        // Assert
+        Assert.Equal(expected, result);
+    }
+
+    [Theory]
+    [InlineData(typeof(float), "number")]
+    [InlineData(typeof(double), "number")]
+    [InlineData(typeof(decimal), "number")]
+    public void ToJsonType_WithNumberTypes_ReturnsNumber(Type type, string expected)
+    {
+        // Act
+        var result = type.ToJsonType();
+
+        // Assert
+        Assert.Equal(expected, result);
+    }
+
+    [Fact]
+    public void ToJsonType_WithBooleanType_ReturnsBoolean()
+    {
+        // Act
+        var result = typeof(bool).ToJsonType();
+
+        // Assert
+        Assert.Equal("boolean", result);
+    }
+
+    [Fact]
+    public void ToJsonType_WithArrayType_ReturnsArray()
+    {
+        // Act
+        var result = typeof(Array).ToJsonType();
+
+        // Assert
+        Assert.Equal("array", result);
+
+        var someArray = new[] { 1, 2, 3 }.GetType();
+
+        Assert.Equal("array", someArray.ToJsonType());
+    }
+
+    [Fact]
+    public void ToJsonType_WithArrayType_ReturnsArray2()
+    {
+        // Act
+        var someArray = new[] { 1, 2, 3 }.GetType();
+
+        // Assert
+        Assert.Equal("array", someArray.ToJsonType());
+    }
+
+
+    [Fact]
+    public void ToJsonType_WithObjectType_ReturnsObject()
+    {
+        // Act
+        var result = typeof(object).ToJsonType();
+
+        // Assert
+        Assert.Equal("object", result);
+    }
+
+    [Theory]
+    [InlineData(typeof(int[]))]
+    [InlineData(typeof(string[]))]
+    [InlineData(typeof(List<int>))]
+    [InlineData(typeof(IList<string>))]
+    [InlineData(typeof(ICollection<object>))]
+    [InlineData(typeof(IEnumerable<int>))]
+    [InlineData(typeof(ArrayList))]
+    public void ToJsonType_WithCollectionTypes_ReturnsArray(Type type)
+    {
+        // Act
+        var result = type.ToJsonType();
+
+        // Assert
+        Assert.Equal("array", result);
+    }
+
+    [Fact]
+    public void ToJsonType_WithStringType_ReturnsString_NotArray()
+    {
+        // Note: string implements IEnumerable<char> but should return "string", not "array"
+        // Act
+        var result = typeof(string).ToJsonType();
+
+        // Assert
+        Assert.Equal("string", result);
+    }
+
+    public enum TestEnum
+    {
+        Value1,
+        Value2
+    }
+
+    [Fact]
+    public void ToJsonType_WithEnumType_ReturnsString()
+    {
+        // Act
+        var result = typeof(TestEnum).ToJsonType();
+
+        // Assert
+        Assert.Equal("string", result);
+    }
+
+    [Theory]
+    [InlineData(typeof(ConsoleColor))]
+    [InlineData(typeof(DayOfWeek))]
+    [InlineData(typeof(FileAttributes))]
+    public void ToJsonType_WithBuiltInEnumTypes_ReturnsString(Type enumType)
+    {
+        // Act
+        var result = enumType.ToJsonType();
+
+        // Assert
+        Assert.Equal("string", result);
+    }
+
+    public class CustomClass
+    {
+        public string Name { get; set; } = string.Empty;
+    }
+
+    public struct CustomStruct
+    {
+        public int Value { get; set; }
+    }
+
+    public interface ICustomInterface
+    {
+        void DoSomething();
+    }
+
+    [Theory]
+    [InlineData(typeof(CustomClass))]
+    [InlineData(typeof(CustomStruct))]
+    [InlineData(typeof(ICustomInterface))]
+    [InlineData(typeof(Exception))]
+    [InlineData(typeof(Stream))]
+    public void ToJsonType_WithCustomTypes_ReturnsObject(Type type)
+    {
+        // Act
+        var result = type.ToJsonType();
+
+        // Assert
+        Assert.Equal("object", result);
+    }
+
+    [Theory]
+    [InlineData(typeof(int?))]
+    [InlineData(typeof(bool?))]
+    [InlineData(typeof(DateTime?))]
+    [InlineData(typeof(TestEnum?))]
+    public void ToJsonType_WithNullableTypes_ReturnsObject(Type nullableType)
+    {
+        // Nullable types are not in the dictionary and don't fall into other categories
+        // Act
+        var result = nullableType.ToJsonType();
+
+        // Assert
+        Assert.Equal("object", result);
+    }
+
+    [Fact]
+    public void ToJsonType_WithGenericType_ReturnsObject()
+    {
+        // Act
+        var result = typeof(Dictionary<string, int>).ToJsonType();
+
+        // Assert
+        Assert.Equal("object", result);
+    }
+
+    [Fact]
+    public void ToJsonType_WithNestedGenericType_ReturnsObject()
+    {
+        // Act
+        var result = typeof(Dictionary<string, List<int>>).ToJsonType();
+
+        // Assert
+        Assert.Equal("object", result);
+    }
+
+    [Fact]
+    public void ToJsonType_WithMultidimensionalArray_ReturnsArray()
+    {
+        // Act
+        var result = typeof(int[,]).ToJsonType();
+
+        // Assert
+        Assert.Equal("array", result);
+    }
+
+    [Fact]
+    public void ToJsonType_WithJaggedArray_ReturnsArray()
+    {
+        // Act
+        var result = typeof(int[][]).ToJsonType();
+
+        // Assert
+        Assert.Equal("array", result);
+    }
+}

--- a/tests/Commands/Server/TypeToJsonTypeMapperTests.cs
+++ b/tests/Commands/Server/TypeToJsonTypeMapperTests.cs
@@ -155,7 +155,7 @@ public class TypeToJsonTypeMapperTests
         var result = typeof(TestEnum).ToJsonType();
 
         // Assert
-        Assert.Equal("string", result);
+        Assert.Equal("integer", result);
     }
 
     [Theory]
@@ -168,7 +168,7 @@ public class TypeToJsonTypeMapperTests
         var result = enumType.ToJsonType();
 
         // Assert
-        Assert.Equal("string", result);
+        Assert.Equal("integer", result);
     }
 
     public class CustomClass

